ICP-7621 Create new dth for the Iris IL071

Automated-House opened the pull request:
The Iris IL071 is similar to the old model, but adds a humidity sensor. This dth is based on the SmartSense motion sensor and adds the humidity pieces if SmartSense Temp/Humidity. Create as a generic zigbee device in case a future device has the same capabilities.

Automated-House committed:
Create zigbee-motion-temp-humidity-sensor.groovy

Automated-House committed:
fix tab

fix humidity capability indention

tpmanley commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,312 @@
+	return configCmds
+}
+
+private shouldUseOldBatteryReporting() {
```
This is specific to Centralite FW so this function and the code that uses it can be removed.

tpmanley commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,312 @@
+	configCmds += zigbee.batteryConfig()
+	configCmds += zigbee.temperatureConfig(30, 300)
+	configCmds += zigbee.readAttribute(zigbee.IAS_ZONE_CLUSTER, zigbee.ATTRIBUTE_IAS_ZONE_STATUS)
+	configCmds += zigbee.readAttribute(zigbee.POWER_CONFIGURATION_CLUSTER, batteryAttr)
```
This can be simplified to just hardcode the appropriate battery attribute (0x0020 or 0x0021 - whichever it supports) and remove the code that looks at the manufacturer

tpmanley commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,312 @@
+
+		command "enrollResponse"
+		
+        //Device join data
```
It's OK to include this information in the commit message but you can remove it from the code since it's basically already covered by the fingerprint itself

tpmanley commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,312 @@
+			input title: "Temperature Offset", description: "This feature allows you to correct any temperature variations by selecting an offset. Ex: If your sensor consistently reports a temp that's 5 degrees too warm, you'd enter '-5'. If 3 degrees too cold, enter '+3'.", displayDuringSetup: false, type: "paragraph", element: "paragraph"
+			input "tempOffset", "number", title: "Degrees", description: "Adjust temperature by this many degrees", range: "*..*", displayDuringSetup: false
+		}
+        section {
```
nit: can you make sure all the indentation is using tabs? I know we haven't been super great about that in the past but we're trying to be more consistent moving forward

tpmanley commented:
Thanks for submitting this @prjct92eh2 ! I made a few minor comments and also plan to try this out at home this weekend

Automated-House committed:
removed old Centralite battery code. removed Samjin battery cluster setting and hard coded the one for this device. removed device info comments.

Automated-House committed:
cleanup spaces/tabs

cleanup spaces/tabs

greens commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,284 @@
+
+		command "enrollResponse"
+		
+        fingerprint inClusters: "0000,0001,0003,0020,0402,0405,0500,0B05,FC01,FC02", outClusters: "0019,0003", manufacturer: "iMagic by GreatStar", model: "1117-S", deviceJoinName: "Iris IL071 Motion Sensor"	
```
missed this one

greens replied:
re: whitespace

greens commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,284 @@
+import physicalgraph.zigbee.zcl.DataType
+
+metadata {
+	definition(name: "Zigbee Motion/Temp/Humidity Sensor", namespace: "smartthings", author: "SmartThings", mnmn: "SmartThings", vid: "generic-motion-6") {
```
`generic-motion-6` also includes the "Illuminance Measurement" capability. That means we'd have an empty tile in the new app, which is generally a no-no. I can see about creating a "generic-motion-8" or similar this weekend, but we don't want to use this one.

Automated-House replied:
yeah, that's the closest I knew of. Let me know if you create a better fit.

tpmanley replied:
I don't see Illuminance Measurement in the new app - did something change?

Automated-House replied:
The new app hides the capability if it’s null.

greens replied:
If that's the case then we don't need to generate new metadata I guess. This would be good to go.

greens commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,284 @@
+		if (humidityOffset) {
+			map.value = (int) map.value + (int) humidityOffset
+		}
+    }
```
more whitespace

your editor probably has a way to convert everything at once if you haven't used that already

Automated-House replied:
yeahhhhh, i haven't been using any editor other than the Groovy IDE and Github. Maybe it's time to put my big boy coding pants on and install one. Any free ones for Windows you would recommend that integrate with Git?

greens replied:
Visual Studio Code is pretty decent (and free) and has great GitHub integration.

Automated-House replied:
@greens I think I got all of these taken care of in the latest commit. VS Code helped!

greens commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,284 @@
+	def refreshCmds = []
+
+
+	refreshCmds += zigbee.readAttribute(zigbee.POWER_CONFIGURATION_CLUSTER, 0x0020) +
```
Crossing my fingers here, but did the device add support for attribute 21? It reports a percentage and saves us the mess of the conversion above.

Automated-House replied:
attribute 20 only I believe.

greens commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,284 @@
+	if (!(rawValue == 0 || rawValue == 255)) {
+		result.name = 'battery'
+		result.translatable = true
+		result.descriptionText = "{{ device.displayName }} battery was {{ value }}%"
```
This doesn't actually result in any string replacement currently in the IDE/app. You should just use the standard groovy `${}`

Automated-House replied:
This was leftover from the SmartSense motion handler. If you think my fixes look correct, I can do the same to the SmartSense motion to fix it.

Automated-House committed:
more cleanup

Automated-House committed:
fixed description text variables, updated configure and refresh

Automated-House committed:
typos and formatting attempts

Automated-House commented:
Let me know how it goes @tpmanley  I think everything is fixed up except some white space wonkiness.

Automated-House committed:
Formatting fixes

fixed formatting with a real editor and cleaned up configure.

greens reviewed (commented):
Looks good, but I want to hold off on putting this in until I get new a new ui-metadata generated and deployed like I mentioned.

Automated-House commented:
sounds good to me!

tpmanley commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,280 @@
+		standardTile("refresh", "device.refresh", inactiveLabel: false, decoration: "flat", width: 2, height: 2) {
+			state "default", action: "refresh.refresh", icon: "st.secondary.refresh"
+		}
+		standardTile("configure", "device.configure", inactiveLabel: false, decoration: "flat", width: 2, height: 2) {
```
We used to have `Configure` tiles in most of our DTHs but we've stopped doing that because the feeling is it confusing to the average customer and shouldn't be needed since configure is called automatically when the device joins.

tpmanley commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,280 @@
+				} else {
+					log.warn "TEMP REPORTING CONFIG FAILED- error code: ${descMap.data[0]}"
+				}
+			} else if (descMap.clusterInt == 0x0406 && descMap.attrInt == 0x0000) {
```
Cluster 0x0406 isn't in the fingerprint. Perhaps this can be removed?

tpmanley commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,280 @@
+		capability "Health Check"
+		capability "Sensor"
+
+		command "enrollResponse"
```
I know we have this is some of our other DTHs but it doesn't serve any purpose anymore that I'm aware of so I think it can be safely removed.

tpmanley commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,280 @@
+	return result
+}
+
+private Map getBatteryPercentageResult(rawValue) {
```
This function isn't used anymore and can be removed

tpmanley commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,280 @@
+
+	configCmds += zigbee.batteryConfig() +
+		zigbee.temperatureConfig(30, 300) +
+		zigbee.configureReporting(0x0405, 0x0000, DataType.UINT16, 30, 3600,100)
```
You can use `zigbee.RELATIVE_HUMIDITY_CLUSTER` instead of 0x0405 here and the other places it's referenced

tpmanley replied:
nit: space before `100`

tpmanley commented on devicetypes/smartthings/zigbee-motion-temp-humidity-sensor.src/zigbee-motion-temp-humidity-sensor.groovy:
```diff
@@ -0,0 +1,280 @@
+		zigbee.temperatureConfig(30, 300) +
+		zigbee.configureReporting(0x0405, 0x0000, DataType.UINT16, 30, 3600,100)
+
+	return configCmds + refresh()
```
I suggest switching the order to decrease the time it takes to get initial status values from the sensor

Automated-House committed:
Update zigbee-motion-temp-humidity-sensor.groovy

removed condigure tile, removed read attr motion event, removed unused battery processing, replaced humidity cluser, swapped refresh and configure order during config call

tpmanley reviewed (approved):
This is working well for me. Just the question about the UI metadata for @greens

greens commented:
Apparently the app is now smart enough to not show capabilities that don't have values, so this works for me.
